feat: ServiceRegistry + Use LocalFileDetector by default

JonasPammer opened the pull request:
Resolves https://github.com/grails/geb/issues/128#issuecomment-2621565934 and takes away my need for a custom utility function and reliance on `container.` as per my current predicament outlined in #134

https://www.selenium.dev/documentation/webdriver/drivers/remote_webdriver/#uploads

JonasPammer committed:
failing test

JonasPammer committed:
statically set as per docs

https://www.selenium.dev/documentation/webdriver/drivers/remote_webdriver/#uploads

JonasPammer committed:
document https://github.com/grails/geb/issues/128#issuecomment-2621565934

JonasPammer commented:
Maybe tomorrow ill make it a new @Shared trait property so it can be overwritten using OOP, like i had previously already too done as replacement for #145 in my codebase https://github.com/grails/geb/blob/4.2.x/src/testFixtures/groovy/grails/plugin/geb/support/ContainerSupport.groovy

jdaugherty commented:
@matrei can you please review this one?

JonasPammer committed:
feat: implement as @ContainerGebConfiguration, add test

TODO factory/serviceloader pattern ability as per https://github.com/grails/geb/pull/143

implementing it as @Shared of ContainerSupport.groovy trait was wrong, and lead to a race condition as not picked up by matchesCurrentContainerConfiguration for obvious reason. also trying that was more complicated to implement, so not even commited to avoid people thinking i have no clue about advanced interlinked self-referencing trait patterns (i dont)

JonasPammer commented:
TODO: implement ServiceLoader Pattern as done in #143 to allow static overwriteability

matrei commented on README.md:
```diff
@@ -76,6 +76,14 @@ By default, no test recording will be performed.  Here are the system properties
 
+#### Uploads
+
+Uploading a file is more complicated for Remote WebDriver sessions because the file you want to upload is likely on the computer executing the code.
```
Suggestion:
... is likely on the host executing the tests and not in the container running the browser.

matrei commented on spock-container-test-app/src/integration-test/groovy/org/demo/spock/LocalUploadSpec.groovy:
```diff
@@ -0,0 +1,48 @@
+        def uploadPage = to UploadPage
+
+        when:
+        uploadPage.fileInput.file = new File("src/integration-test/resources/assets/upload-test.txt")
```
Single quotes for Strings?

matrei commented on spock-container-test-app/src/integration-test/groovy/org/demo/spock/LocalUploadSpec.groovy:
```diff
@@ -0,0 +1,48 @@
+        def uploadPage = to UploadPage
+
+        when:
+        uploadPage.fileInput.file = new File("src/integration-test/resources/assets/upload-test.txt")
```
Single quotes for Strings?

matrei commented on spock-container-test-app/src/integration-test/groovy/org/demo/spock/UselessUploadSpec.groovy:
```diff
@@ -0,0 +1,50 @@
+        def uploadPage = to UploadPage
+
+        when:
+        uploadPage.fileInput.file = new File("src/integration-test/resources/assets/upload-test.txt")
```
Single quotes for Strings?

matrei commented on spock-container-test-app/src/integration-test/groovy/org/demo/spock/UselessUploadSpec.groovy:
```diff
@@ -0,0 +1,50 @@
+
+        then:
+        def e = thrown(WebDriverException)
+        e.message.contains("File not found")
```
Single quotes for Strings?

matrei commented on spock-container-test-app/src/integration-test/groovy/org/demo/spock/UselessUploadSpec.groovy:
```diff
@@ -0,0 +1,50 @@
+        def uploadPage = to UploadPage
+
+        when:
+        uploadPage.fileInput.file = new File("src/integration-test/resources/assets/upload-test.txt")
```
Single quotes for Strings?

matrei commented on spock-container-test-app/src/integration-test/groovy/org/demo/spock/UselessUploadSpec.groovy:
```diff
@@ -0,0 +1,50 @@
+
+        then:
+        def e = thrown(WebDriverException)
+        e.message.contains("File not found")
```
Single quotes for Strings?

matrei commented on src/testFixtures/groovy/grails/plugin/geb/ContainerGebConfiguration.groovy:
```diff
@@ -52,4 +56,19 @@ import java.lang.annotation.Target
+     *
+     * <p> Must have a zero-argument constructor.
+     *
+     * @since 4.1
```
`@since 4.2`?

matrei commented on src/testFixtures/groovy/grails/plugin/geb/ContainerGebConfiguration.groovy:
```diff
@@ -52,4 +56,19 @@ import java.lang.annotation.Target
+     * @since 4.1
+     * @see org.openqa.selenium.remote.LocalFileDetector LocalFileDetector (grails.geb default)
+     * @see org.openqa.selenium.remote.UselessFileDetector UselessFileDetector (null/unset value)
+     * @see
```
Empty `@see`

matrei commented on src/testFixtures/groovy/grails/plugin/geb/ContainerGebConfiguration.groovy:
```diff
@@ -52,4 +56,19 @@ import java.lang.annotation.Target
+     * @see org.openqa.selenium.remote.UselessFileDetector UselessFileDetector (null/unset value)
+     * @see
+     */
+    @NotNull(message="You may not set a file detector that is null.")
```
This is not necessary and I don't think it works as intended anyway:
```console
Only classes and closures can be used for attribute 'fileDetector' in @grails.plugin.geb.ContainerGebConfiguration
 @ line 16, column 43.
   bConfiguration(fileDetector = null)
                                 ^
```

matrei reviewed (commented):
This is excellent! Now users don’t have to do anything special for uploads to work seamlessly in `ContainerGebSpec`.

Looking forward to your addition of a `ServiceLoader` implementation.

JonasPammer commented:
probably heading to do this and others on weekend

> Looking forward to your addition of a ServiceLoader implementation.

Since its my first time using this pattern, do you think it'd be ok if we create just one Loader Class for everything (which i cant implement here, as multiple PRs) or one Loader for each corresponding trait/interface (the latter of which i am going to write here)
no-op not-important question, just curious

in the github code search projects i couldnt find much and if, it was always seperate small price utility loader class for one thing. whole idea of this question is i dont want to bloat the code with dozens of loaders in future

matrei commented:
Good point! Now that we have created our own Loader with override feature in #143, we could make that more generic (for the class) and specific for the methods. Then we can add loading of different types in that class as separate PR:s after #143 is merged.

An alternative would be to create multiple loaders grouped in a package like `grails.geb.serviceloader.*`.
I think I actually like this alternative better. Then we get these classes out of the main package and can still have sensible and aligned method names (`setInstance()`).

JonasPammer committed:
chore: code review

JonasPammer committed:
try to replicate serviceloaderity

more minimalist serviceloader/spock/annotation README

made a interface to allow future extensibility, i.e. maybe pass along current container info through an interface getter, similiar to how we pass errorInfo in other serviceloader?
may have went overboard with this

JonasPammer committed:
Merge branch '4.2.x' into file-support

JonasPammer commented:
may have went overboard, please correct me if simpler better or mistakes

matrei commented on README.md:
```diff
@@ -76,6 +76,25 @@ By default, no test recording will be performed.  Here are the system properties
+   on the classpath (e.g., `src/integration-test/resources`).
+2. Call [`ContainerFileDetectorServiceLoader.setInstance()`](./src/testFixtures/groovy/grails/plugin/geb/serviceloader/ContainerFileDetectorServiceLoader.groovy)
+   in a Spock `setupSpec` method to apply your naming convention (And use a `cleanupSpec` to limit this to one class).
+3. Set the `fileDetector` Property on your `ContainerGebConfiguration` annotation.
```
Suggestion:
Use the `ContainerGebConfiguration` annotation and set its `fileDetector` property to your `ContainerFileDetector` implementation class.

matrei commented on README.md:
```diff
@@ -76,6 +76,25 @@ By default, no test recording will be performed.  Here are the system properties
+   and specify its fully qualified class name in a `META-INF/services/grails.plugin.geb.ContainerFileDetector` file
+   on the classpath (e.g., `src/integration-test/resources`).
+2. Call [`ContainerFileDetectorServiceLoader.setInstance()`](./src/testFixtures/groovy/grails/plugin/geb/serviceloader/ContainerFileDetectorServiceLoader.groovy)
+   in a Spock `setupSpec` method to apply your naming convention (And use a `cleanupSpec` to limit this to one class).
```
`setupSpec()` `cleanupSpec()`?

matrei commented on README.md:
```diff
@@ -76,6 +76,25 @@ By default, no test recording will be performed.  Here are the system properties
+3. Set the `fileDetector` Property on your `ContainerGebConfiguration` annotation.
+
+Alternatively, you can access the `BrowserWebDriverContainer` instance via
+the `container` from within your ContainerGebSpec to for example call `.copyFileToContainer`,
```
Backticks for `ContainerGebSpec`?
Parentheses on method name to make it clearer that it is a method: ... to for example call one of its `copyFile*()` methods ...

matrei commented on README.md:
```diff
@@ -76,6 +76,25 @@ By default, no test recording will be performed.  Here are the system properties
+the `container` from within your ContainerGebSpec to for example call `.copyFileToContainer`,
+as done in [ContainerSupport#createFileInputSource utility method](./src/testFixtures/groovy/grails/plugin/geb/support/ContainerSupport.groovy).
+
 #### Remove Implicit Wait
```
This heading is actually a bit off now. I know it was merged from another PR, but maybe we can change it here to get it done in 4.2.x?
Suggestion:
Timeouts

matrei commented on src/testFixtures/groovy/grails/plugin/geb/ContainerFileDetector.groovy:
```diff
@@ -0,0 +1,32 @@
+ */
+@CompileStatic
+interface ContainerFileDetector extends FileDetector {
+    
```
Remove empty line?

matrei commented on src/testFixtures/groovy/grails/plugin/geb/ContainerGebConfiguration.groovy:
There are some unused imports in this file.

matrei commented on src/testFixtures/groovy/grails/plugin/geb/DefaultContainerFileDetector.groovy:
```diff
@@ -0,0 +1,45 @@
+/**
+ * Used by {@link ContainerGebConfiguration#fileDetector()} interface to represent a null value.
+ */
+class NullContainerFileDetector extends DefaultContainerFileDetector {}
```
Align empty class space handling.

matrei commented on src/testFixtures/groovy/grails/plugin/geb/WebDriverContainerHolder.groovy:
There are some unused imports in this file.

matrei commented:
How about we create base service registry:
```groovy
package grails.plugin.geb

import groovy.transform.CompileStatic

import java.util.concurrent.ConcurrentHashMap

/**
 * A service registry that loads and caches different service types using the Java
 * {@link java.util.ServiceLoader}, while allowing overriding which instance to return.
 * <p>
 * This class provides thread-safe service loading and caching, supporting parallel test execution.
 * It provides both automatic service discovery through {@link java.util.ServiceLoader}
 * and explicit replacement of the returned service instance for customization.
 * </p>
 * <p>
 * Usage example:
 * <pre>
 * MyService service = ServiceRegistry.getInstance(MyService, DefaultMyService)
 * </pre>
 * </p>
 *
 * @since 4.2
 * @author Mattias Reichel
 */
@CompileStatic
class ServiceRegistry {

    private static final ThreadLocal<ConcurrentHashMap<Class<?>, Object>> INSTANCES = ThreadLocal.withInitial {
        new ConcurrentHashMap<Class<?>, Object>()
    }

    /**
     * Returns the service instance of the given service type, loading it using {@link java.util.ServiceLoader} if not
     * already loaded or returning the default instance if no service implementation is found.
     *
     * If an instance has been set using {@link #setInstance(Class, Object)} or {@link #setInstance(Class, Class)}
     * that instance will be returned instead.
     *
     * @param serviceType The service type
     * @param defaultClass The service implementation type to use if no service implementation is found
     * (Must have a zero-argument constructor)
     * @return The service instance
     */
    static <T> T getInstance(Class<T> serviceType, Class<? extends T> defaultClass) {
        (T) INSTANCES.get().computeIfAbsent(serviceType) {
            ServiceLoader.load(serviceType)
                    .findFirst()
                    .orElseGet { defaultClass.getDeclaredConstructor().newInstance() }
        }
    }

    /**
     * Sets the instance to be returned for the given service type, bypassing instance loading
     * from {@link java.util.ServiceLoader}.
     * <p>
     * Setting the instance to {@code null} will revert to loading the the service instance via the
     * {@link java.util.ServiceLoader}.
     * </p>
     * @param serviceType The service type for which the instance should be set
     * @param instance The instance to return for the given service type, or {@code null} for default service loading
     */
    static <T> void setInstance(Class<T> serviceType, T instance) {
        INSTANCES.get().put(serviceType, instance)
    }

    /**
     * Sets the implementation type to return for the given service type, bypassing instance loading
     * from {@link java.util.ServiceLoader}.
     *
     * @param serviceType The service type for which the instance should be set
     * @param instanceType The type of the instance to return for the given service type.
     * (Must have a zero-argument constructor).
     */
    static <T> void setInstance(Class<T> serviceType, Class<? extends T> instanceType) {
        setInstance(serviceType, instanceType.getDeclaredConstructor().newInstance())
    }
}
```

then we can delegate to it and do not have to manage any implementation details in the different concrete implementations:
```groovy
@CompileStatic
class ContainerFileDetectorServiceLoader {

    /**
     * Delegates to {@link ServiceRegistry#getInstance(Class, Class)}.
     */
    static ContainerFileDetector getInstance() {
        ServiceRegistry.getInstance(ContainerFileDetector, DefaultContainerFileDetector)
    }

    /**
     * Delegates to {@link ServiceRegistry#setInstance(Class, Object)}.
     */
    static void setInstance(ContainerFileDetector instance) {
        ServiceRegistry.setInstance(ContainerFileDetector, instance)
    }

    /**
     * Delegates to {@link ServiceRegistry#setInstance(Class, Class)}.
     */
    static void setInstance(Class<? extends ContainerFileDetector> clazz) {
        ServiceRegistry.setInstance(ContainerFileDetector, clazz)
    }
}
```

or perhaps even cleaner, call the `ServiceRegistry ` directly not needing any different service loader implementations at all:
```groovy
if (currentConfiguration.fileDetector != NullContainerFileDetector) {
    ServiceRegistry.setInstance(ContainerFileDetector, currentConfiguration.fileDetector)
}
ContainerFileDetector fileDetector = ServiceRegistry.getInstance(ContainerFileDetector, DefaultContainerFileDetector)
```

What do you think?

JonasPammer commented:
I knew about threadlocal, to be honest only through a random claude generation once, but yes with this concurrenthashmap it looks like solid code. Can you tell me why we need this threadlocal or concurrent thing though? If not ill study myself. But yes, totally can do with this

matrei commented:
As the  `Loader`  class is used statically (same `instance` for all threads), when running tests in parallel (different threads, in the same JVM process), multiple threads can set different service implementations at the same time, you can't know that the instance that is meant to be used in a test is actually used, it could have been set by another thread. With `ThreadLocal` each thread gets its own version of the service registry map. Hmm, maybe it does not have to be a `ConcurrentHashMap`, I had that as a thread safety guard first, but then realized that would not suffice as different threads could still replace each others service instances.

jdaugherty commented:
> As the `Loader` class is used statically (same `instance` for all threads), when running tests in parallel (different threads, in the same JVM process), multiple threads can set different service implementations at the same time, you can't know that the instance that is meant to be used in a test is actually used, it could have been set by another thread. With `ThreadLocal` each thread gets its own version of the service registry map. Hmm, maybe it does not have to be a `ConcurrentHashMap`, I had that as a thread safety guard first, but then realized that would not suffice as different threads could still replace each others service instances.

Usually `Context` type classes use a thread local to ensure isolation.  I don't think we need to optimize to share these across threads - the overhead isn't that extreme.  Let's change to a ThreadLocal and ditch the hashmap?

matrei commented:
Actually, `TheadLocal` might be overkill as well: https://github.com/grails/geb?tab=readme-ov-file#parallel-execution.

JonasPammer commented:
> > Actually, `TheadLocal` might be overkill as well: https://github.com/grails/geb?tab=readme-ov-file#parallel-execution.
> 
> Long term, I was hoping to make it thread safe.  I did this as a stop gap for now.  I'd rather we go ahead and add threadlocal for future proofing.

Totally, hell yeah

JonasPammer committed:
preliminary code review

JonasPammer commented:
put to draft until i get to doing wished implementation of comments

JonasPammer committed:
copy/paste https://github.com/grails/geb/pull/146#issuecomment-2690523380

did not (yet) implement the third part of mentioned comment. still using ContainerFileDetectorServiceLoader

only difference: put into .serviceloader package

JonasPammer committed:
copy/paste "call the ServiceRegistry directly not needing any different service loader implementations at all"

JonasPammer committed:
replace the concurrenthashmap with hashmap (https://github.com/grails/geb/pull/146#issuecomment-2691320778)

JonasPammer committed:
rename spec classes to fit with #143

JonasPammer committed:
test confirming https://github.com/grails/geb/pull/146#issuecomment-2691433277 (no exception was thrown)

JonasPammer commented:
> `WebdriverContainerHolder.reinitialize()` seems to run before `setupSpec()` so the instance will already be set on the driver when `setupSpec()` runs:
> 
> ```groovy
> ContainerFileDetector fileDetector = ContainerFileDetectorServiceLoader.getInstance()
> ((RemoteWebDriver) driver).setFileDetector(fileDetector)
> ```

Yes, some added println with flushes confirm:
```
matchesCurrentContainerConfiguration grails.plugin.geb.WebDriverContainerHolder$WebDriverContainerConfiguration(http, host.testcontainers.internal, false, class grails.plugin.geb.DefaultContainerFileDetector) == grails.plugin.geb.WebDriverContainerHolder$WebDriverContainerConfiguration(http, host.testcontainers.internal, false, class grails.plugin.geb.UselessContainerFileDetector)
2025-03-04 07:38:27.867  INFO 137172 --- [    Test worker] g.plugin.geb.WebDriverContainerHolder    : ContainerFileDetectorDefaultSpec does not match grails.plugin.geb.UselessContainerFileDetector@5b2a23e9
matchesCurrentContainerConfiguration grails.plugin.geb.WebDriverContainerHolder$WebDriverContainerConfiguration(http, host.testcontainers.internal, false, class grails.plugin.geb.DefaultContainerFileDetector) == grails.plugin.geb.WebDriverContainerHolder$WebDriverContainerConfiguration(http, host.testcontainers.internal, false, class grails.plugin.geb.DefaultContainerFileDetector)


2025-03-04 07:38:36.297  INFO 137172 --- [    Test worker] g.plugin.geb.WebDriverContainerHolder    : ContainerFileDetectorSpockSpec matches grails.plugin.geb.DefaultContainerFileDetector@7c68b975
matchesCurrentContainerConfiguration grails.plugin.geb.WebDriverContainerHolder$WebDriverContainerConfiguration(http, host.testcontainers.internal, false, class grails.plugin.geb.DefaultContainerFileDetector) == grails.plugin.geb.WebDriverContainerHolder$WebDriverContainerConfiguration(http, host.testcontainers.internal, false, class grails.plugin.geb.DefaultContainerFileDetector)
ContainerFileDetectorSpockSpec::setupSpec


2025-03-04 07:38:36.730  INFO 137172 --- [    Test worker] g.plugin.geb.WebDriverContainerHolder    : DownloadSupportSpec matches grails.plugin.geb.DefaultContainerFileDetector@328c9bf5
matchesCurrentContainerConfiguration grails.plugin.geb.WebDriverContainerHolder$WebDriverContainerConfiguration(http, host.testcontainers.internal, false, class grails.plugin.geb.DefaultContainerFileDetector) == grails.plugin.geb.WebDriverContainerHolder$WebDriverContainerConfiguration(http, host.testcontainers.internal, false, class grails.plugin.geb.DefaultContainerFileDetector)
2025-03-04 07:38:37.528  INFO 137172 --- [    Test worker] g.plugin.geb.WebDriverContainerHolder    : PageDelegateSpec matches grails.plugin.geb.DefaultContainerFileDetector@328c9bf5
matchesCurrentContainerConfiguration grails.plugin.geb.WebDriverContainerHolder$WebDriverContainerConfiguration(http, host.testcontainers.internal, true, class grails.plugin.geb.DefaultContainerFileDetector) == grails.plugin.geb.WebDriverContainerHolder$WebDriverContainerConfiguration(http, host.testcontainers.internal, false, class grails.plugin.geb.DefaultContainerFileDetector)
2025-03-04 07:38:37.751  INFO 137172 --- [    Test worker] g.plugin.geb.WebDriverContainerHolder    : RootPageSpec does not match grails.plugin.geb.DefaultContainerFileDetector@328c9bf5
matchesCurrentContainerConfiguration grails.plugin.geb.WebDriverContainerHolder$WebDriverContainerConfiguration(http, testing.example.com, false, class grails.plugin.geb.DefaultContainerFileDetector) == grails.plugin.geb.WebDriverContainerHolder$WebDriverContainerConfiguration(http, host.testcontainers.internal, true, class grails.plugin.geb.DefaultContainerFileDetector)
2025-03-04 07:38:45.389  INFO 137172 --- [    Test worker] g.plugin.geb.WebDriverContainerHolder    : ServerNameControllerSpec does not match grails.plugin.geb.DefaultContainerFileDetector@382b008d
matchesCurrentContainerConfiguration grails.plugin.geb.WebDriverContainerHolder$WebDriverContainerConfiguration(http, host.testcontainers.internal, false, class grails.plugin.geb.DefaultContainerFileDetector) == grails.plugin.geb.WebDriverContainerHolder$WebDriverContainerConfiguration(http, testing.example.com, false, class grails.plugin.geb.DefaultContainerFileDetector)
2025-03-04 07:38:52.992  INFO 137172 --- [    Test worker] g.plugin.geb.WebDriverContainerHolder    : UploadSpec does not match grails.plugin.geb.DefaultContainerFileDetector@437eeb4

matchesCurrentContainerConfiguration grails.plugin.geb.WebDriverContainerHolder$WebDriverContainerConfiguration(http, host.testcontainers.internal, false, class grails.plugin.geb.DefaultContainerFileDetector) == currentConfiguration
2025-03-04 07:37:24.922  INFO 134180 --- [    Test worker] g.plugin.geb.WebDriverContainerHolder    : ContainerFileDetectorSpockSpec matches grails.plugin.geb.DefaultContainerFileDetector@2c5ff4a5
ContainerFileDetectorSpockSpec::setupSpec
```
Until figure how to circumvent, will stay draft

JonasPammer commented:
> If Spock is run in parallell mode, setupSpec() will probably not run in the same thread as the feature methods, so anything we set there will not be set in the feature methods.

To be honest, this is a bit too complicated for me to resolve quickly, as for once i have never used parallel mode, and for twice not even #135 could manage similiar affair of the chicken and the egg. The two functionalities to implement an overwrite are already good as they are, given than 95% won't even bother to need it.

To make it better than that, I will however follow up with an PR to implement an `inherited()` in the ContainerGebConfiguration Annotation, just because i just saw it in [Spock](https://spockframework.org/spock/javadoc/2.3/spock/lang/Ignore.html#inherited()). This I think will resolve the intent of the Spock methodology.

Even if this ThreadLocal doesn't work as envisioned, i don't think it will hurt, so ill let that stay i think.

JonasPammer committed:
Ignore SpockSpec with GitHub URL, Update docs

matrei commented:
Thanks @JonasPammer!  I'm have a look at this as soon as a get a chance.

JonasPammer commented:
Thank you for heads up, very kind. I show my apprentices at work your review work to teach them about Git Merge Requests :) .

matrei commented on spock-container-test-app/src/integration-test/groovy/org/demo/spock/ContainerFileDetectorAnnotationSpec.groovy:
```diff
@@ -0,0 +1,51 @@
+class ContainerFileDetectorAnnotationSpec extends ContainerGebSpec {
+
+    @Requires({ os.windows })
+    void 'should be able to find and upload files on a Windows host'() {
```
Can we adjust this feature method title to better match what is tested?
Suggestion: `should fail to find file with fileDetector changed to UselessContainerFileDetector via annotation`

Also, do we need to test different operating systems?

matrei commented on spock-container-test-app/src/integration-test/groovy/org/demo/spock/ContainerFileDetectorDefaultSpec.groovy:
Do we need to test different operating systems here?

matrei commented on spock-container-test-app/src/integration-test/groovy/org/demo/spock/ContainerFileDetectorSpockSpec.groovy:
```diff
@@ -0,0 +1,61 @@
+    }
+    
+    @Requires({ os.windows })
+    void 'should be able to find and upload files on a Windows host'() {
```
Can we adjust this feature method title to better match what is tested?
Suggestion: `should fail to find file with fileDetector changed to UselessContainerFileDetector in setupSpec`

matrei commented on spock-container-test-app/src/integration-test/groovy/org/demo/spock/ContainerFileDetectorSpockSpec.groovy:
```diff
@@ -0,0 +1,61 @@
+ * Altered copy of {@link ContainerFileDetectorAnnotationSpec}
+ */
+@Integration
+@Ignore("https://github.com/grails/geb/pull/146#issuecomment-2691433277")
```
Use `@PendingFeature(reason = 'https://github.com/grails/geb/pull/146#issuecomment-2691433277')` on the feature method instead of `@Ignore`?

Do we need to test different operating systems here?

Unused imports?

matrei commented on src/testFixtures/groovy/grails/plugin/geb/ContainerGebConfiguration.groovy:
Update year range in license header?

matrei commented on src/testFixtures/groovy/grails/plugin/geb/DefaultContainerFileDetector.groovy:
```diff
@@ -0,0 +1,44 @@
+}
+
+/**
+ * Used by {@link ContainerGebConfiguration#fileDetector()} interface to represent a null value.
```
Should this class also be documented with `@since 4.2`?

matrei commented on src/testFixtures/groovy/grails/plugin/geb/WebDriverContainerHolder.groovy:
```diff
@@ -118,10 +118,14 @@ class WebDriverContainerHolder {
             configObject.reporter = (invocation.sharedInstance as ContainerGebSpec).createReporter()
         }
+        if (currentConfiguration.fileDetector != NullContainerFileDetector) {
+            ServiceRegistry.setInstance(ContainerFileDetector, currentConfiguration.fileDetector)        }
```
Spacing?

matrei commented on src/testFixtures/groovy/grails/plugin/geb/serviceloader/ServiceRegistry.groovy:
```diff
@@ -0,0 +1,79 @@
+package grails.plugin.geb.serviceloader
+import groovy.transform.CompileStatic
+
+import java.util.concurrent.ConcurrentHashMap
```
Unused import

matrei commented on src/testFixtures/groovy/grails/plugin/geb/serviceloader/ServiceRegistry.groovy:
```diff
@@ -0,0 +1,79 @@
+package grails.plugin.geb.serviceloader
+import groovy.transform.CompileStatic
```
Line break between `package` and first `import`?

matrei commented on src/testFixtures/groovy/grails/plugin/geb/serviceloader/ServiceRegistry.groovy:
Should we have a license header in this file?

matrei reviewed (changes requested):
Well done! Some minor comments...

JonasPammer committed:
reviewed

tests works on my windows machine

matrei commented on spock-container-test-app/src/integration-test/groovy/org/demo/spock/ContainerFileDetectorAnnotationSpec.groovy:
Unused imports

matrei commented on spock-container-test-app/src/integration-test/groovy/org/demo/spock/ContainerFileDetectorDefaultSpec.groovy:
Unused imports

matrei commented on src/testFixtures/groovy/grails/plugin/geb/DefaultContainerFileDetector.groovy:
Any reason we have `@CompileStatic` on 2 of 3 classes?
Do we actually need it on empty classes?

JonasPammer replied:
no, and claude (always-agreeing ai, so duh) also says no) thanks

JonasPammer committed:
resolve

matrei commented on src/testFixtures/groovy/grails/plugin/geb/DefaultContainerFileDetector.groovy:
Unused import 😉 ?

JonasPammer replied:
thanks, currently have deactivated most autosave or styling as doing refactor on other big project

JonasPammer committed:
intellij optimize imports

JonasPammer commented:
## unrelated euphoria

Very cool to see this project having moved to apache! I originally figured the rename from grails/geb to grails/grails-geb was related to the decision of https://github.com/apache/grails-core/discussions/13516. Is there a thread for how the integration with apache came to be, out of interest?

edit: wait, digging deeper i just realised ["everything"](https://github.com/orgs/apache/repositories?type=source&q=grails) moved to apache...
edit: found at least a spec - https://groovy.apache.org/wiki/grails-proposal.html
        does the move in github means it got accepted ([this hints to a yes](https://grails.org/blog/2025-03-05-grails-7-m3.html))?

JonasPammer commented:
> This is excellent! Now users don’t have to do anything special for uploads to work seamlessly in `ContainerGebSpec`.

Just a bump for this PR, would be cool to see this in 7.0's first release :). 
I of course accept the new Apache ICLA and will send it today
